File: programmers/hash/충돌위험찾기.py
from collections import defaultdict

def solution(points, routes):
    pos_points = {i + 1: point for i, point in enumerate(points)}
    timedict = defaultdict(int)
    
    for route in routes:
        time = 0
        len_route = len(route)
        r, c = pos_points[route[0]]
        timedict[(time, r, c)] += 1

        for i in range(1, len_route):
            nr, nc = pos_points[route[i]]
            while r != nr:
                r += 1 if r < nr else -1
                time += 1
                timedict[(time, r, c)] += 1

            while c != nc:
                c += 1 if c < nc else -1
                time += 1
                timedict[(time, r, c)] += 1

    danger = sum(1 for cnt in timedict.values() if cnt >= 2)
    
    return danger

points = [[3, 2], [6, 4], [4, 7], [1, 4]]
routes = [[4, 2], [1, 3], [2, 4]]
print(solution(points, routes))

File: programmers/graph/충돌위험찾기.js
function solution(points, routes) {
    var answer = 0;
    const map = new Map();
    
    for (const route of routes) {
        let [cur_row, cur_col] = points[route[0] - 1];
        let time = 0;
        const key = [cur_row, cur_col, time].toString(); 
        map.set(key, (map.get(key) || 0) + 1);
        
        for (let i = 1; i < route.length; i++) {
            const [next_row, next_col] = points[route[i] - 1];
            
            while (cur_row !== next_row) {
                if (cur_row < next_row) cur_row++;
                else cur_row--;
                const key = [cur_row, cur_col, ++time].toString();
                map.set(key, (map.get(key) || 0) + 1);
            }
            
            while (cur_col !== next_col) {
                if (cur_col < next_col) cur_col++;
                else cur_col--;
                const key = [cur_row, cur_col, ++time].toString();
                map.set(key, (map.get(key) || 0) + 1);
            }
        }
    }
    for (const v of map.values()) {
        if (v > 1) answer++;
    }

    return answer;
}